sycl/include/sycl/ext/intel/experimental/fpga_mem/fpga_mem.hpp

```diff
@@ -0,0 +1,138 @@
+//==----------- fpga_mem.hpp - SYCL fpga_mem extension -----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/access/access.hpp> // for address_space
+#include <sycl/exception.hpp>     // for make_error_code
+#include <sycl/ext/intel/experimental/fpga_mem/properties.hpp> // for num_banks
+#include <sycl/ext/oneapi/properties/properties.hpp> // for properties_t
+// #include <sycl/pointers.hpp>                         // for decorated_gl...
+
+#include <cstddef>     // for ptrdiff_t
+#include <type_traits> // for enable_if_t
+#include <utility>     // for declval
+
+namespace sycl {
+inline namespace _V1 {
+namespace ext::intel::experimental {
+// shorthand used with properties
+namespace oneapi_exp = ext::oneapi::experimental;
+namespace detail {
+
+// Hide the base implementation in details so that manipulation
+// of properties parameter pack can be modularized away from main logic
+template <typename T, typename... Props> class fpga_mem_base {
+
+protected:
+  T val
+#ifdef __SYCL_DEVICE_ONLY__
+      // In addition to annotating all the user specified properties, also add
+      // {"sycl-resource", ""} hinting to the optimizer that this object
+      // should be implemented in memory outside the datapath.
+      [[__sycl_detail__::add_ir_annotations_member(
+          "sycl-resource", oneapi_exp::detail::PropertyMetaInfo<Props>::name...,
+          "DEFAULT", oneapi_exp::detail::PropertyMetaInfo<Props>::value...)]]
+#endif
+      ;
+
+  T *get_ptr() noexcept { return &val; }
+  constexpr const T *get_ptr() const noexcept { return &val; }
```

Comment:
I have a feeling this was just adopted from what `device_global` does, but is there actually a need got these? Could the `get()` methods not just return a reference to `val` directly?